internal/repository/index.go

```diff
@@ -26,27 +28,72 @@ type Index struct {
 }
 
 type indexEntry struct {
-	packID restic.ID
-	offset uint
-	length uint
+	packID *restic.ID
+	offset uint32
+	length uint32
 }
 
 // NewIndex returns a new index.
 func NewIndex() *Index {
 	return &Index{
-		pack:    make(map[restic.BlobHandle][]indexEntry),
-		created: time.Now(),
+		blob:       make(map[restic.BlobHandle]indexEntry),
+		duplicates: make(map[restic.BlobHandle][]indexEntry),
+		created:    time.Now(),
 	}
 }
 
-func (idx *Index) store(blob restic.PackedBlob) {
+// allBlobs return all blobs to the given blobHandle saved in the index
+// also returns if there exists a blob to the handle or not.
+func (idx *Index) allBlobs(h restic.BlobHandle) ([]indexEntry, bool) {
+	blob, ok := idx.blob[h]
+	if !ok {
+		return nil, false
+	}
+
+	entries, okDup := idx.duplicates[h]
+	if okDup {
+		all := make([]indexEntry, len(entries)+1)
+		all[0] = blob
+		copy(all[1:], entries)
+		return all, true
+	}
+
+	return []indexEntry{blob}, true
+}
+
+// get Pack gets the pointer to the given pack ID in the array of saved pack IDs
+// it ID is not present, saves it in that array and returns the pointer to the
+// new entry.
+// This procedere allows to use pack IDs which can be easily garbage collected after.
+func (idx *Index) getPack(id restic.ID) *restic.ID {
+	for _, packID := range idx.packs {
+		if id == packID {
+			return &packID
```

Comment:
This doesn't work as intended. The loop variable is a copy, so taking its address copies it to the heap:

```go
package main

import "fmt"

func main() {
	strs := []string{"foo"}

	var p *string
	for _, s := range strs {
		p = &s
	}

	*p = "bar"
	fmt.Println(strs) // [foo], not [bar]
}
```

You need to use an index if you want a pointer into the slice.

Reply:
Oh thanks a lot! Now you say it I recall that I read this copy-on-loop somewhere, just forgot about it..
Is now changed.